File: apina-core/src/main/java/fi/evident/apina/java/reader/TypeParser.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.MethodSignature;
import fi.evident.apina.java.model.type.JavaArrayType;
import fi.evident.apina.java.model.type.JavaBasicType;
import fi.evident.apina.java.model.type.JavaType;
import fi.evident.apina.java.model.type.TypeSchema;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.objectweb.asm.Type;
import org.objectweb.asm.signature.SignatureReader;

import java.util.List;

import static fi.evident.apina.utils.CollectionUtils.cons;
import static fi.evident.apina.utils.CollectionUtils.map;

/**
 * Parses Java's internal type names, descriptors and signatures into
 */
final class TypeParser {

    public static JavaType parseJavaType(@NotNull String typeDescriptor, @Nullable String signature) {
        if (signature != null)
            return parseGenericType(signature);
        else
            return parseTypeDescriptor(typeDescriptor);
    }

    public static JavaType parseGenericType(String signature) {
        TypeBuildingSignatureVisitor visitor = new TypeBuildingSignatureVisitor();

        new SignatureReader(signature).acceptType(visitor);

        return visitor.get();
    }

    public static JavaType parseTypeDescriptor(String typeDescriptor) {
        return javaType(Type.getType(typeDescriptor));
    }

    public static JavaBasicType parseBasicTypeDescriptor(String typeDescriptor) {
        return parseTypeDescriptor(typeDescriptor).toBasicType();
    }

    public static JavaType parseObjectType(String internalName) {
        return javaType(Type.getObjectType(internalName));
    }

    public static MethodSignature parseMethodSignature(String methodDescriptor, @Nullable String signature) {
        MethodSignature legacySignature = parseMethodDescriptor(methodDescriptor);
        if (signature != null) {
            MethodSignature genericSignature = parseGenericMethodSignature(signature);

            // There are classes in the wild with differing argument counts for the two signatures.
            // At least guava-jdk5-13.0's com/google/common/collect/AbstractMultimap$WrappedSortedSet
            // is like this. All of the cases are the same: the implicit parent-class argument as the
            // first argument is left out. If we detect that, create a new argument list mostly based
            // on the generic signature, but using the parent type from non-generic descriptor.
            if (legacySignature.getArgumentCount() == genericSignature.getArgumentCount() + 1) {
                JavaType implicitParentType = legacySignature.getArgumentTypes().get(0);
                return new MethodSignature(
                        genericSignature.getReturnType(),
                        cons(implicitParentType, genericSignature.getArgumentTypes()),
                        genericSignature.getSchema());
            } else {
                assert legacySignature.getArgumentCount() == genericSignature.getArgumentCount();
                return genericSignature;
            }
        }

        return legacySignature;
    }

    static MethodSignature parseGenericMethodSignature(String signature) {
        MethodSignatureVisitor visitor = new MethodSignatureVisitor();

        new SignatureReader(signature).accept(visitor);

        return visitor.get();
    }

    @NotNull
    static MethodSignature parseMethodDescriptor(String methodDescriptor) {
        Type methodType = Type.getMethodType(methodDescriptor);

        JavaType returnType = javaType(methodType.getReturnType());
        List<JavaType> argumentTypes = map(methodType.getArgumentTypes(), TypeParser::javaType);

        return new MethodSignature(returnType, argumentTypes, new TypeSchema());
    }

    private static JavaType javaType(Type type) {
        if (type.getSort() == Type.ARRAY) {
            JavaType javaType = new JavaBasicType(type.getElementType().getClassName());

            for (int i = 0, dimensions = type.getDimensions(); i < dimensions; i++)
                javaType = new JavaArrayType(javaType);

            return javaType;

        } else {
            return new JavaBasicType(type.getClassName());
        }
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/reader/TypeParser.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.MethodSignature
import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeSchema
import org.objectweb.asm.Type
import org.objectweb.asm.signature.SignatureReader

fun parseJavaType(typeDescriptor: String, signature: String?): JavaType =
    if (signature != null)
        parseGenericType(signature)
    else
        parseTypeDescriptor(typeDescriptor)

fun parseGenericType(signature: String): JavaType {
    val visitor = TypeBuildingSignatureVisitor()

    SignatureReader(signature).acceptType(visitor)

    return visitor.get()
}

fun parseTypeDescriptor(typeDescriptor: String) = javaType(Type.getType(typeDescriptor))

fun parseBasicTypeDescriptor(typeDescriptor: String): JavaType.Basic = parseTypeDescriptor(typeDescriptor).toBasicType()

fun parseObjectType(internalName: String) = javaType(Type.getObjectType(internalName))

fun parseMethodSignature(methodDescriptor: String, signature: String?): MethodSignature {
    val legacySignature = parseMethodDescriptor(methodDescriptor)
    if (signature != null) {
        val genericSignature = parseGenericMethodSignature(signature)

        // There are classes in the wild with differing argument counts for the two signatures.
        // At least guava-jdk5-13.0's com/google/common/collect/AbstractMultimap$WrappedSortedSet
        // is like this. All of the cases are the same: the implicit parent-class argument as the
        // first argument is left out. If we detect that, create a new argument list mostly based
        // on the generic signature, but using the parent type from non-generic descriptor.
        return if (legacySignature.argumentCount == genericSignature.argumentCount + 1) {
            val implicitParentType = legacySignature.argumentTypes[0]
            MethodSignature(
                genericSignature.returnType,
                listOf(implicitParentType) + genericSignature.argumentTypes,
                genericSignature.schema
            )
        } else {
            assert(legacySignature.argumentCount == genericSignature.argumentCount)
            genericSignature
        }
    }

    return legacySignature
}

fun parseGenericMethodSignature(signature: String): MethodSignature {
    val visitor = MethodSignatureVisitor()

    SignatureReader(signature).accept(visitor)

    return visitor.get()
}

fun parseMethodDescriptor(methodDescriptor: String): MethodSignature {
    val methodType = Type.getMethodType(methodDescriptor)

    val returnType = javaType(methodType.returnType)
    val argumentTypes = methodType.argumentTypes.map(::javaType)

    return MethodSignature(returnType, argumentTypes, TypeSchema())
}

fun javaType(type: Type): JavaType =
    if (type.sort == Type.ARRAY) {
        var javaType: JavaType = JavaType.Basic(type.elementType.className)

        repeat(type.dimensions) {
            javaType = JavaType.Array(javaType)
        }

        javaType

    } else {
        JavaType.Basic(type.className)
    }
